airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py

```diff
@@ -0,0 +1,167 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+
+import uuid
+from dataclasses import dataclass
+from typing import Any, Dict, Iterable, Mapping, Optional
+
+import requests
+from airbyte_cdk.models import FailureType
+from airbyte_cdk.sources.declarative.async_job.job import AsyncJob
+from airbyte_cdk.sources.declarative.async_job.repository import AsyncJobRepository
+from airbyte_cdk.sources.declarative.async_job.status import AsyncJobStatus
+from airbyte_cdk.sources.declarative.extractors.dpath_extractor import DpathExtractor, RecordExtractor
+from airbyte_cdk.sources.declarative.extractors.response_to_file_extractor import ResponseToFileExtractor
+from airbyte_cdk.sources.declarative.requesters.requester import Requester
+from airbyte_cdk.sources.types import StreamSlice
+from airbyte_cdk.utils import AirbyteTracedException
+from requests import Response
+
+
+@dataclass
+class AsyncHttpJobRepository(AsyncJobRepository):
+    creation_requester: Requester
+    polling_requester: Requester
+    download_requester: Requester
+    status_extractor: DpathExtractor
+    status_mapping: Mapping[str, AsyncJobStatus]
+    urls_extractor: DpathExtractor
+
+    record_extractor: RecordExtractor = ResponseToFileExtractor()
+
+    def __post_init__(self) -> None:
+        self._create_job_response_by_id: Dict[str, Response] = {}
+        self._polling_job_response_by_id: Dict[str, Response] = {}
+
+    def _get_validated_polling_response(self, stream_slice: StreamSlice) -> requests.Response:
+        """
+        Validates and retrieves the pooling response for a given stream slice.
+
+        Args:
+            stream_slice (StreamSlice): The stream slice to send the pooling request for.
+
+        Returns:
+            requests.Response: The validated pooling response.
+
+        Raises:
+            AirbyteTracedException: If the polling request returns an empty response.
+        """
+
+        polling_response: Optional[requests.Response] = self.polling_requester.send_request(stream_slice=stream_slice)
+        if polling_response is None:
+            raise AirbyteTracedException(
+                internal_message="Polling Requester received an empty Response.",
+                failure_type=FailureType.system_error,
+            )
+        return polling_response
+
+    def _get_validated_job_status(self, response: requests.Response) -> AsyncJobStatus:
+        """
+        Validates the job status extracted from the API response.
+
+        Args:
+            response (requests.Response): The API response.
+
+        Returns:
+            AsyncJobStatus: The validated job status.
+
+        Raises:
+            ValueError: If the API status is unknown.
+        """
+
+        api_status = next(iter(self.status_extractor.extract_records(response)), None)
+        job_status = self.status_mapping.get(str(api_status), None)
+        if job_status is None:
+            raise ValueError(
```

Comment:
I'm wondering if this is dangerous: what happens if the API adds a status? I guess we should know in some way but Salesforce today only operate on a COMPLETE and FAILED statuses and probably assume the rest is running (see [this](https://github.com/airbytehq/airbyte/blob/f23ae56693c931e36028a61ea1b11ffdd25762e2/airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py#L421))

Reply:
yes, this will fail eventually, but I think the desired behavior is indeed to fail loudly if an unexpected status shows up. Happy to revisit the decision if this becomes too noisy and isn't actionable.

Reply:
I'm afraid of this creating P0s as we don't control the API but I also get your point. Let's try and learn from that